generator/metalines_builder.cpp

```diff
@@ -0,0 +1,239 @@
+#include "generator/metalines_builder.hpp"
+#include "generator/routing_helpers.hpp"
+
+#include "indexer/classificator.hpp"
+
+#include "coding/file_container.hpp"
+#include "coding/varint.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "base/checked_cast.hpp"
+#include "base/logging.hpp"
+#include "base/string_utils.hpp"
+
+#include "defines.hpp"
+
+#include <cstdint>
+
+namespace
+{
+uint8_t const kMetaLinesSectionVersion = 1;
+
+using Ways = std::vector<int32_t>;
+
+/// A string of connected ways.
+class LineString
+{
+  Ways m_ways;
+  uint64_t m_start;
+  uint64_t m_end;
+  bool m_oneway;
+
+public:
+  explicit LineString(OsmElement const & way)
+  {
+    std::string const oneway = way.GetTag("oneway");
+    m_oneway = !oneway.empty() && oneway != "no";
+    int32_t const wayId = base::checked_cast<int32_t>(way.id);
+    m_ways.push_back(oneway == "-1" ? -wayId : wayId);
+    CHECK_GREATER_OR_EQUAL(way.Nodes().size(), 2, ());
+    m_start = way.Nodes().front();
+    m_end = way.Nodes().back();
+  }
+
+  Ways const & Ways() const { return m_ways; }
+
+  void Reverse()
+  {
+    ASSERT(!m_oneway, ("Trying to reverse a one-way road."));
+    std::swap(m_start, m_end);
+    std::reverse(m_ways.begin(), m_ways.end());
+    for (auto & p : m_ways)
+      p = -p;
+  }
+
+  bool Add(LineString & line)
+  {
+    if (m_start == line.m_start || m_end == line.m_end)
+    {
+      if (!line.m_oneway)
+        line.Reverse();
+      else if (!m_oneway)
+        Reverse();
+      else
+        return false;
+    }
+    if (m_end == line.m_start)
+    {
+      m_ways.insert(m_ways.end(), line.m_ways.begin(), line.m_ways.end());
+      m_end = line.m_end;
+      m_oneway = m_oneway || line.m_oneway;
+    }
+    else if (m_start == line.m_end)
+    {
+      m_ways.insert(m_ways.begin(), line.m_ways.begin(), line.m_ways.end());
+      m_start = line.m_start;
+      m_oneway = m_oneway || line.m_oneway;
+    }
+    else
+    {
+      return false;
+    }
+    return true;
+  }
+};
+}  // namespace
+
+namespace feature
+{
+/// A list of segments, that is, LineStrings, sharing the same attributes.
+class Segments
+{
+  std::list<LineString> m_parts;
+
+public:
+  explicit Segments(OsmElement const & way) { m_parts.emplace_back(way); }
+
+  void Add(OsmElement const & way)
+  {
+    LineString line(way);
+    auto found = m_parts.end();
+    for (auto i = m_parts.begin(); i != m_parts.end(); ++i)
+    {
+      if (i->Add(line))
+      {
+        found = i;
+        break;
+      }
+    }
+    if (found == m_parts.cend())
+    {
+      m_parts.push_back(line);
+      return;
+    }
+    for (LineString & part : m_parts)
+    {
+      if (part.Add(*found))
```

Comment:
Could you please explain why do we need to add found to some element of m_parts?